### DISCUSSION TITLE
Connecting to SQL database from Vercel functions

### DISCUSSION THREAD

**Discussion by DavidChouinard** [vercel/vercel#8802] [Help] [answered]
What's the best way to connect to an SQL database from Vercel/Next.js functions? So far I've seen these options commonly suggested:

- **Connect directly to database from the serverless function**. This has 2 important drawbacks:
  - Connection pooling: every new function invocation would create a new database connection, potentially overwhelming the database
  - Security: the database would have to be publicly exposed, which unfortunately is a non-started in regulated industries (finance, healthcare, etc.) and for SOC 2 compliance
- **Use an intermediate service**, something that would accept queries over HTTP and proxy the request to the database. How do people set this up?
- **Use a vendor-specific solution**, like [Prisma Data Proxy](https://www.prisma.io/docs/data-platform/data-proxy) (requires using the Prisma ORM) or AWS Aurora Data API (now deprecated)

Trying to understand what the "best practice" is for this problem — any good solution that people have been happy with?

**Comment by amyegan** [comment]
It looks like you have already been researching this, but you should check out the guide on [Using Databases with Vercel](https://vercel.com/guides/using-databases-with-vercel)! Caching or connection pooling via an HTTP API are common solutions.
> Rather than opening a connection with every request, connection pooling allows us to designate a single "pooler" that keeps an active connection to the database. When a user invokes a Serverless Function that would read from the database, it will first search for an already available connection instead of creating a new connection.

Using a service that provides an HTTP API for database interaction is the most straightforward approach, through I'm sure others may have come up with various workarounds. There is a list of available [database integrations](https://vercel.com/integrations#databases). They are designed to make data connection easy with providers optimized for serverless environments.

**Comment by DavidChouinard** [comment, marked as answer]
Ah I see. Unfortunately we're stuck with a traditional SQL database. It sounds like the best option then as adding a layer on top with an HTTP API?

**Reply by amyegan** [reply]
Yeah, I think that's your best bet to avoid migrating the database to a different service
